core/src/View/Kokkos_ViewCommonType.hpp

```diff
@@ -0,0 +1,128 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_VIEWCOMMONTYPE_HPP
+#define KOKKOS_VIEWCOMMONTYPE_HPP
+
+#include <type_traits>
+#include <Kokkos_Macros.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+template <class Specialize, typename A, typename B>
+struct CommonViewValueType;
+
+template <typename A, typename B>
+struct CommonViewValueType<void, A, B> {
+  using value_type = std::common_type_t<A, B>;
+};
+
+template <class Specialize, class ValueType>
+struct CommonViewAllocProp;
+
+template <class ValueType>
+struct CommonViewAllocProp<void, ValueType> {
+  using value_type        = ValueType;
+  using scalar_array_type = ValueType;
+
+  template <class... Views>
+  KOKKOS_INLINE_FUNCTION CommonViewAllocProp(const Views&...) {}
+};
+
+template <class... Views>
+struct DeduceCommonViewAllocProp;
+
+// Base case must provide types for:
+// 1. specialize  2. value_type  3. is_view  4. prop_type
+template <class FirstView>
+struct DeduceCommonViewAllocProp<FirstView> {
+  using specialize = typename FirstView::traits::specialize;
+
+  using value_type = typename FirstView::traits::value_type;
+
+  enum : bool { is_view = is_view<FirstView>::value };
+
+  using prop_type = CommonViewAllocProp<specialize, value_type>;
+};
+
+template <class FirstView, class... NextViews>
+struct DeduceCommonViewAllocProp<FirstView, NextViews...> {
+  using NextTraits = DeduceCommonViewAllocProp<NextViews...>;
+
+  using first_specialize = typename FirstView::traits::specialize;
+  using first_value_type = typename FirstView::traits::value_type;
+
+  enum : bool { first_is_view = is_view<FirstView>::value };
+
+  using next_specialize = typename NextTraits::specialize;
+  using next_value_type = typename NextTraits::value_type;
+
+  enum : bool { next_is_view = NextTraits::is_view };
+
+  // common types
+
+  // determine specialize type
+  // if first and next specialize differ, but are not the same specialize, error
+  // out
+  static_assert(!(!std::is_same_v<first_specialize, next_specialize> &&
+                  !std::is_void_v<first_specialize> &&
+                  !std::is_void_v<next_specialize>),
```

Comment:
This can be simplified to
```c++
static_assert(std::is_same_v<first_specialize, next_specialize> || std::is_void_v<first_specialize> || std::is_void_v<next_specialize>));
```

Reply:
Also just moving code here not sure if I want to drive-by fix this